src/cargo/sources/registry/index.rs

```diff
@@ -99,51 +99,81 @@ impl<'cfg> RegistryIndex<'cfg> {
             .collect::<String>();
 
         // see module comment for why this is structured the way it is
-        let path = match fs_name.len() {
+        let raw_path = match fs_name.len() {
             1 => format!("1/{}", fs_name),
             2 => format!("2/{}", fs_name),
             3 => format!("3/{}/{}", &fs_name[..1], fs_name),
             _ => format!("{}/{}/{}", &fs_name[0..2], &fs_name[2..4], fs_name),
         };
+        let num_hyphen_underscore = ::std::cmp::min(
+            raw_path.chars().filter(|&c| c == '_' || c == '-').count(),
+            10,
+        ) as u16;
         let mut ret = Vec::new();
-        let mut hit_closure = false;
-        let err = load.load(&root, Path::new(&path), &mut |contents| {
-            hit_closure = true;
-            let contents = str::from_utf8(contents)
-                .map_err(|_| format_err!("registry index file was not valid utf-8"))?;
-            ret.reserve(contents.lines().count());
-            let lines = contents.lines().map(|s| s.trim()).filter(|l| !l.is_empty());
-
-            let online = !self.config.cli_unstable().offline;
-            // Attempt forwards-compatibility on the index by ignoring
-            // everything that we ourselves don't understand, that should
-            // allow future cargo implementations to break the
-            // interpretation of each line here and older cargo will simply
-            // ignore the new lines.
-            ret.extend(lines.filter_map(|line| {
-                let (summary, locked) = match self.parse_registry_package(line) {
-                    Ok(p) => p,
-                    Err(e) => {
-                        info!("failed to parse `{}` registry package: {}", name, e);
-                        trace!("line: {}", line);
-                        return None;
+        // Crates.io treats hyphen and underscores as interchangeable
+        // but, the index and old cargo do not. So the index must store uncanonicalized version
+        // of the name so old cargos can find it.
+        // This loop tries all possible combinations of switching
+        // hyphen and underscores to find the uncanonicalized one.
+        for hyphen_combination_num in 0u16..(1 << num_hyphen_underscore) {
```

Comment:
This might be a pretty good candidate to refactor out into a custom iterator, and that way we could just do `.take(1000)` to avoid too many permutations from happening?

Reply:
The min 10 above already limits this, but an additional `.take(1024)` to make it absolutely clear cant hurt. I'll look into making a dedicated iterator. And possibly having a smaller cut off with a fallback to `glob`.

Reply:
Oh sure yeah what I mean is it'll be functionally equivalent but perhaps a bit cleaner on the implementation

Reply:
Moved to a struct. It allowed adding more targeted tests!